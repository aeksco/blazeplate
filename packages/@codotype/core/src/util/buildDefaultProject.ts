import {
    SchemaInput,
    ProjectInput,
    PropertyType,
    OptionValue,
    PluginConfiguration,
    OptionValueInstance,
    ConfigurationGroup,
    ConfigurationGroupSection,
    PluginMetadata,
    ConfigurationProperty,
} from "../";

// // // //

/**
 * buildConfigurationPropertyValue
 * Accepts a single ConfigurationProperty and recursively produces its associated OptionValue instance
 * @param property - the single ConfigurationProperty for which the OptionValue is being produced
 */
export function buildConfigurationPropertyValue(
    property: ConfigurationProperty,
): OptionValue {
    if (property.type === PropertyType.STRING) {
        return property.defaultValue || "";
    }
    if (property.type === PropertyType.BOOLEAN) {
        return property.defaultValue || false;
    }
    if (property.type === PropertyType.DROPDOWN) {
        return property.defaultValue
            ? property.defaultValue
            : property.dropdownOptions.length > 0
            ? property.dropdownOptions[0].value
            : "";
    }
    if (property.type === PropertyType.COLLECTION) {
        if (Array.isArray(property.defaultValue)) {
            return property.defaultValue;
        }
        // Return empty array as default
        return [];
    }
    if (property.type === PropertyType.INSTANCE) {
        return buildValueFromProperties(property.properties);
    }
    return "";
}

/**
 * buildValueFromProperties
 * Builds the top-level OptionValueInstance for a single ConfigurationGroup
 * @param properties - array of ConfigurationProperty
 */
export function buildValueFromProperties(
    properties: ConfigurationProperty[],
): OptionValueInstance {
    const initialValue: OptionValueInstance = {};

    // Defines empty ConfigurationGroupValue
    // Iterates over each property in the group and assigns values
    const configurationGroupValue: OptionValueInstance = properties.reduce(
        (val, property: ConfigurationProperty) => {
            // Updates val with data for ConfigurationProperty
            if (property.allowDisable && !property.required) {
                val[property.identifier] = {
                    enabled: property.enabledByDefault,
                    value: buildConfigurationPropertyValue(property),
                };
            } else {
                val[property.identifier] = buildConfigurationPropertyValue(
                    property,
                );
            }
            // Returns val
            return val;
        },
        initialValue, // Passes in empty OptionValueInstance
    );

    // Returns ConfigurationGroupValue
    return configurationGroupValue;
}

/**
 * buildConfigurationGroupValue
 * Builds the top-level OptionValueInstance for a single ConfigurationGroup
 * @param properties - array of ConfigurationProperty
 */
export function buildConfigurationGroupValue(
    configurationGroup: ConfigurationGroup,
): OptionValueInstance {
    const initialValue: OptionValueInstance = {};

    // Handle ConfigurationGroup w/ properties
    if (configurationGroup.properties) {
        // Defines empty ConfigurationGroupValue
        // Iterates over each property in the group and assigns values
        const configurationGroupValue: OptionValueInstance = configurationGroup.properties.reduce(
            (val, property: ConfigurationProperty) => {
                // Updates val with data for ConfigurationProperty
                // TODO - remove this - the UI should only care if the property can allow disable or not
                // TODO - remove this - the UI should only care if the property can allow disable or not
                // TODO - remove this - the UI should only care if the property can allow disable or not
                // TODO - remove this - the UI should only care if the property can allow disable or not
                if (property.allowDisable && !property.required) {
                    val[property.identifier] = {
                        enabled: property.enabledByDefault,
                        value: buildConfigurationPropertyValue(property),
                    };
                } else {
                    val[property.identifier] = buildConfigurationPropertyValue(
                        property,
                    );
                }
                // Returns val
                return val;
            },
            initialValue, // Passes in empty OptionValueInstance
        );

        // Returns ConfigurationGroupValue
        return configurationGroupValue;
    }

    // Handle ConfigurationGroup w/ Sections
    if (configurationGroup.sections) {
        // Iterate over each section
        const configurationGroupValue: OptionValueInstance = configurationGroup.sections.reduce(
            (val, section: ConfigurationGroupSection) => {
                // Defines initial value for the ConfigurationGroupSection
                const initialSectionValue: OptionValueInstance = {};

                // Iterates over each property in the ConfigurationGroupSection
                const sectionValue: OptionValueInstance = section.properties.reduce(
                    (val, property: ConfigurationProperty) => {
                        // Sets value for ConfigurationGroupSection
                        val[property.identifier] = {
                            enabled: property.enabledByDefault,
                            value: buildConfigurationPropertyValue(property),
                        };

                        // Returns val
                        return val;
                    },
                    initialSectionValue, // Passes in empty OptionValueInstance
                );

                // Returns ConfigurationGroupValue
                return {
                    ...initialValue,
                    [section.identifier]: sectionValue,
                };
            },
            initialValue,
        );

        // Returns the configurationGroupValue
        return configurationGroupValue;
    }
}

/**
 * buildDefaultConfiguration
 * Builds a new PluginConfiguration instance based on an array of ConfigurationGroup instances
 * @param configurationGroups - array of ConfigurationGroup instances
 */
export function buildDefaultConfiguration(
    configurationGroups: ConfigurationGroup[],
): PluginConfiguration {
    // Defines default PluginConfiguration
    const PluginConfiguration: PluginConfiguration = configurationGroups.reduce(
        (val, configurationGroup: ConfigurationGroup) => {
            const initialValue: OptionValueInstance = buildConfigurationGroupValue(
                configurationGroup,
            );
            return { ...val, [configurationGroup.identifier]: initialValue };
        },
        {}, // Passes in empty PluginConfiguration
    );

    return PluginConfiguration;
}

/**
 * buildDefaultProject
 * Builds an empty Project
 * @param pluginMetadata - the PluginMetadata for-which the ProjectInput is being constructed
 */
export function buildDefaultProjectInput(
    pluginMetadata: PluginMetadata,
): ProjectInput {
    // Defines default PluginConfiguration
    const PluginConfiguration: PluginConfiguration = buildDefaultConfiguration(
        pluginMetadata.configurationGroups,
    );

    // Returns ConfigurationGroupValue
    const projectInput: ProjectInput = {
        id: "",
        identifiers: {
            title: "New Project",
            snake: "new_project",
            camel: "newProject",
            pascal: "NewProject",
            kebab: "new-project",
        },
        pluginID: pluginMetadata.identifier,
        pluginVersion: pluginMetadata.version,
        schemas: [...pluginMetadata.schemaEditorConfiguration.defaultSchemas],
        relations: [
            ...pluginMetadata.schemaEditorConfiguration.defaultRelations,
        ],
        configuration: PluginConfiguration,
    };

    // Returns the new project
    return projectInput;
}
