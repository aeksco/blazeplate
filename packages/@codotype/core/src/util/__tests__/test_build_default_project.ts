import { buildDefaultProjectInput } from "../buildDefaultProject";
import {
    testState,
    PluginMetadata,
    ProjectInput,
    CreatedByValues,
} from "../../";

// // // //

const testCases: [string, PluginMetadata][] = [
    ["cdkGeneratorMeta", testState.cdkGeneratorMeta],
    ["dummyGeneratorMeta", testState.dummyGeneratorMeta],
    [
        "generator w/ defaultSchemas, no schemaEditorConfiguration.configurationGroups",
        {
            ...testState.dummyGeneratorMeta,
            schemaEditorConfiguration: {
                ...testState.dummyGeneratorMeta.schemaEditorConfiguration,
                defaultSchemas: [
                    {
                        ...testState.userSchema,
                        locked: true,
                        createdBy: CreatedByValues.plugin,
                    },
                ],
                configurationGroups: [],
            },
        },
    ],
    [
        "generator w/ defaultSchemas + schemaEditorConfiguration.configurationGroups",
        {
            ...testState.dummyGeneratorMeta,
            schemaEditorConfiguration: {
                ...testState.dummyGeneratorMeta.schemaEditorConfiguration,
                defaultSchemas: [
                    {
                        ...testState.userSchema,
                        locked: true,
                        createdBy: CreatedByValues.plugin,
                    },
                ],
                configurationGroups: [testState.ApiExamplesConfigurationGroup],
            },
        },
    ],
];

describe("/buildDefaultProject", () => {
    testCases.forEach((testCase) => {
        test(testCase[0], () => {
            const expectedProject: ProjectInput = buildDefaultProjectInput(
                testCase[1],
            );
            expect(expectedProject).toMatchSnapshot();
        });
    });
});
