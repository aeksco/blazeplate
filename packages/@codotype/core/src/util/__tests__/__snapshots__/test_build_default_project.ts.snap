// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/buildDefaultProject cdkGeneratorMeta 1`] = `
Object {
  "configuration": Object {
    "api_examples": Object {
      "twitter": false,
    },
    "component_builder_side_by_side": Object {
      "twitter": false,
    },
    "components_group": Object {
      "componentName": "",
      "dropdownTest": "OPTION_01",
      "numberTest": "",
      "tests": Object {
        "enabled": true,
        "value": Object {
          "componentName": "",
          "dropdownTest": "OPTION_01",
          "nested_instance": Object {
            "enabled": true,
            "value": Object {
              "componentName": "",
              "dropdownTest": "OPTION_01",
              "random_instance": Object {
                "enabled": true,
                "value": Object {
                  "dropdownTest": "OPTION_01",
                },
              },
            },
          },
        },
      },
    },
    "lambda_builder": Object {
      "lambda_collection": Object {
        "enabled": true,
        "value": Array [],
      },
    },
  },
  "id": "",
  "identifiers": Object {
    "camel": "newProject",
    "kebab": "new-project",
    "label": "New Project",
    "pascal": "NewProject",
    "snake": "new_project",
  },
  "pluginID": "aws_cdk_starter",
  "pluginVersion": "0.1.0",
  "schemas": Array [],
}
`;

exports[`/buildDefaultProject dummyGeneratorMeta 1`] = `
Object {
  "configuration": Object {
    "api_examples": Object {
      "aviary": false,
      "chartjs": false,
      "clockwork_sms": false,
      "facebook": false,
      "foursquare": false,
      "github": false,
      "google_maps": false,
      "instagram": false,
      "last_fm": true,
      "linkedin": true,
      "lob": false,
      "paypal": false,
      "pinterest": false,
      "steam": false,
      "stripe": false,
      "tumblr": false,
      "twilio": false,
      "twitter": false,
      "web_scraping": false,
    },
    "component_builder_side_by_side": Object {
      "twitter": false,
    },
    "components_group": Object {
      "componentName": "",
      "dropdownTest": "OPTION_01",
      "numberTest": "",
      "tests": Object {
        "enabled": true,
        "value": Object {
          "componentName": "",
          "dropdownTest": "OPTION_01",
          "nested_instance": Object {
            "enabled": true,
            "value": Object {
              "componentName": "",
              "dropdownTest": "OPTION_01",
              "random_instance": Object {
                "enabled": true,
                "value": Object {
                  "dropdownTest": "OPTION_01",
                },
              },
            },
          },
        },
      },
    },
  },
  "id": "",
  "identifiers": Object {
    "camel": "newProject",
    "kebab": "new-project",
    "label": "New Project",
    "pascal": "NewProject",
    "snake": "new_project",
  },
  "pluginID": "chrome_extension_generator",
  "pluginVersion": "0.1.0",
  "schemas": Array [],
}
`;

exports[`/buildDefaultProject generator w/ defaultSchemas + schemaEditorConfiguration.configurationGroups 1`] = `
Object {
  "configuration": Object {
    "api_examples": Object {
      "aviary": false,
      "chartjs": false,
      "clockwork_sms": false,
      "facebook": false,
      "foursquare": false,
      "github": false,
      "google_maps": false,
      "instagram": false,
      "last_fm": true,
      "linkedin": true,
      "lob": false,
      "paypal": false,
      "pinterest": false,
      "steam": false,
      "stripe": false,
      "tumblr": false,
      "twilio": false,
      "twitter": false,
      "web_scraping": false,
    },
    "component_builder_side_by_side": Object {
      "twitter": false,
    },
    "components_group": Object {
      "componentName": "",
      "dropdownTest": "OPTION_01",
      "numberTest": "",
      "tests": Object {
        "enabled": true,
        "value": Object {
          "componentName": "",
          "dropdownTest": "OPTION_01",
          "nested_instance": Object {
            "enabled": true,
            "value": Object {
              "componentName": "",
              "dropdownTest": "OPTION_01",
              "random_instance": Object {
                "enabled": true,
                "value": Object {
                  "dropdownTest": "OPTION_01",
                },
              },
            },
          },
        },
      },
    },
  },
  "id": "",
  "identifiers": Object {
    "camel": "newProject",
    "kebab": "new-project",
    "label": "New Project",
    "pascal": "NewProject",
    "snake": "new_project",
  },
  "pluginID": "chrome_extension_generator",
  "pluginVersion": "0.1.0",
  "schemas": Array [
    Object {
      "attributes": Array [],
      "configuration": Object {
        "api_examples": Object {
          "twitter": false,
        },
      },
      "id": "12345",
      "identifiers": Object {
        "plural": Object {
          "camel": "users",
          "kebab": "users",
          "label": "Users",
          "pascal": "Users",
          "snake": "users",
        },
        "singular": Object {
          "camel": "user",
          "kebab": "user",
          "label": "User",
          "pascal": "User",
          "snake": "user",
        },
      },
      "internalNote": "",
      "locked": true,
      "removable": false,
      "source": "plugin",
    },
  ],
}
`;

exports[`/buildDefaultProject generator w/ defaultSchemas, no schemaEditorConfiguration.configurationGroups 1`] = `
Object {
  "configuration": Object {
    "api_examples": Object {
      "aviary": false,
      "chartjs": false,
      "clockwork_sms": false,
      "facebook": false,
      "foursquare": false,
      "github": false,
      "google_maps": false,
      "instagram": false,
      "last_fm": true,
      "linkedin": true,
      "lob": false,
      "paypal": false,
      "pinterest": false,
      "steam": false,
      "stripe": false,
      "tumblr": false,
      "twilio": false,
      "twitter": false,
      "web_scraping": false,
    },
    "component_builder_side_by_side": Object {
      "twitter": false,
    },
    "components_group": Object {
      "componentName": "",
      "dropdownTest": "OPTION_01",
      "numberTest": "",
      "tests": Object {
        "enabled": true,
        "value": Object {
          "componentName": "",
          "dropdownTest": "OPTION_01",
          "nested_instance": Object {
            "enabled": true,
            "value": Object {
              "componentName": "",
              "dropdownTest": "OPTION_01",
              "random_instance": Object {
                "enabled": true,
                "value": Object {
                  "dropdownTest": "OPTION_01",
                },
              },
            },
          },
        },
      },
    },
  },
  "id": "",
  "identifiers": Object {
    "camel": "newProject",
    "kebab": "new-project",
    "label": "New Project",
    "pascal": "NewProject",
    "snake": "new_project",
  },
  "pluginID": "chrome_extension_generator",
  "pluginVersion": "0.1.0",
  "schemas": Array [
    Object {
      "attributes": Array [],
      "configuration": Object {},
      "id": "12345",
      "identifiers": Object {
        "plural": Object {
          "camel": "users",
          "kebab": "users",
          "label": "Users",
          "pascal": "Users",
          "snake": "users",
        },
        "singular": Object {
          "camel": "user",
          "kebab": "user",
          "label": "User",
          "pascal": "User",
          "snake": "user",
        },
      },
      "internalNote": "",
      "locked": true,
      "removable": false,
      "source": "plugin",
    },
  ],
}
`;
